### DISCUSSION TITLE
Micropython development on MacOS for STM Controllers

### DISCUSSION THREAD

**Discussion by WolleKette** [micropython/micropython#11238] [STM32 / Pyboard]
So far I have developed software for ESP32 controllers under MacOS. Now I need a controller with 12-bit ADC and 12-bit DAC for a project and came across the controllers from STM (e. g. STM32L452xx). I read that these are supported by Micropython and was excited. 
But I find confusing information about the toolchain on macOS.

There I find instructions how to install a STM32CubeIDE under MacOS and somewhere else it says that MacOS is not supported by STM32CubeIDE. Then there is STLINK, supposedly the predecessor of STM32CubeIDE, at Homebrew available for the Mac.

Does anyone have experience with using MacOS (Intel) to develop micropython software for STM controllers?

**Comment by dlech** [comment]
MicroPython includes it's own copy of the STM32 Cube libraries in a git submodule, so there is nothing special about macOS in this regard. You do not need any special software other than the Arm GCC toolchain.

**Reply by WolleKette** [reply]
Ok, I understand.
But one thing is unclear to me. Yesterday I experimented with the STM32L432KC. As a trial I connected the module to Windows PC. And after installing the driver from STM, I could connect via USB / REPL. Per Thonny or putty for example.
On the Mac, I see a serial device /dev/cu.... from STM, but I can't connect. Neither via screen nor Thonny/VS code etc.

It is clear that the STM32L432KC board has too less flash memory for a file system. But if the access via REPL is possible from Windows, then it should also work from MacOS, right? And I have not found any driver etc. from STM for macOS.

**Reply by dlech** [reply]
I think I usually use the `/dev/tty*` device rather than the `/dev/cu*` device. There are also tools like `mpremote` that simplify things.

**Reply by WolleKette** [reply]
I've now ordered a NUCLEO-L476RG board. 
With mpremote, I get a repl from my mac. And nicely, the board is even supported a bit by PyCharm. There I also get a REPL connection. But I'm unable to connect via VS-Code or Thonny. 

In my other projects with ESP32 boards, I have always written the programs in PyCharm (comfortable editor) and transferred and debugged them via Thonny. I have also experimented with VS code. But with the STM boards I have no idea how to work with it.

**Reply by WolleKette** [reply]
I would like to add something after playing around with it a bit. The whole thing is quite unstable. 
There are statements that verifiably lead to a crash. dir(pyb.Pin) for example.

```
>>> dir(pyb.Pin)
['__class__', '__name__', 'dict', 'value', '__bases__', '__dict__', 'AF12_SDMMC1', 'AF14_TIM15', 'AF14_TIM16', 'AF14_TIM17', 'AF14_TIM2', 'AF1_TIM1', 'AF1_TIM2', 'AF2_TIM2', 'AF2_TIM3', 'AF2_TIM4', 'AF2_TIM5', 'AF3_TIM8', 'AF4_I2C1', 'AF4_I2C2', 'AF4_I2C3', 'AF5_SPI1', 'AF5_SPI2', 'AF7_USART1', 'AF7_USART2', 'AF7_USART3', 'AF8_LPUART1', 'AF8_UART4', 'AF8_UART5', 'AF9_CAN1', 'AF_OD', 'AF_PP', 'ALT', 'ALT_OPEN_DRAIN', 'ANALOG', 'IN', 'IRQ_FALLING', 'IRQ_RISING', 'OPEN_DRAIN', 'OUT', 'OUT_OD', 'OUT_PP', 'PULL_DOWN', 'PULL_NONE', 'PULL_UP', 'af', 'af_list', 'board', 'cpu', 'debug', 'gpio', 'high', 'init', 'irq', 'low', 'mapper', 'mode', 'name', 'names', 'off', 'on', 'pin', 'port', 'pull']

PROBLEM IN THONNY'S BACK-END: Exception while handling 'execute_source' (thonny.plugins.micropython.mp_back.ProtocolError: Could not read command confirmation).
See Thonny's backend.log for more info.
You may need to press "Stop/Restart" or hard-reset your MicroPython device and try again.


Process ended with exit code 1.
```

I can't find anything special in the mentioned log files

**Reply by robert-hh** [reply]
Try that without Thonny, e.g
 using mpremote for REPL.

**Reply by WolleKette** [reply]
Oh sorry, the post accidentally ended up in the wrong forum. I had also reported the problem in the Thonny Support Forum and got a hint there. But that doesn't matter. Maybe it also interests someone here.

Before I could connect with Thonny and consequently before my attempts with the dir() function (my post above), I had to make one more change in  the Thonny configuration file. So if someone has the same problem, here is the link to the thread: 
[Thonny - Backend problem with Nucleo L476RG](https://groups.google.com/g/thonny/c/znltkEc7l1s)
